### DISCUSSION TITLE
How to use `tower_http::normalize_path` with TestServer?

### DISCUSSION THREAD

**Discussion by agersant** [tokio-rs/axum#2833] [Q&A] [answered]
### Summary

Hello! I was able to make use of the tower `normalize path` layer thanks to the tips I found in https://github.com/tokio-rs/axum/discussions/2377.

However, I can't seem to find a way to create a Axum [TestServer](https://docs.rs/axum-test/latest/axum_test/struct.TestServer.html) that also applies path normalization.

This code snippet:

```rust
let router = Router::new();
let app = NormalizePathLayer::trim_trailing_slash().layer(router);
let app = ServiceExt::<axum::extract::Request>::into_make_service(app);

let server = TestServer::new(app).unwrap();
```

 fails to compile with the following error:
 
 ```
error[E0277]: the trait bound `axum::routing::IntoMakeService<tower_http::normalize_path::NormalizePath<axum::Router>>: axum_test::transport_layer::IntoTransportLayer` is not satisfied
  --> src\server\axum\test.rs:40:32
   |
40 |         let server = TestServer::new(app).unwrap();
   |                      --------------- ^^^ the trait `axum_test::transport_layer::IntoTransportLayer` is not implemented for `axum::routing::IntoMakeService<tower_http::normalize_path::NormalizePath<axum::Router>>`
   |                      |
   |                      required by a bound introduced by this call
   |
   = help: the trait `axum_test::transport_layer::IntoTransportLayer` is implemented for `axum::routing::IntoMakeService<axum::Router>
```

I understand what the error message is saying but I'm not sure what to do about it. Any help would be welcome!

### axum version

0.7.5

**Comment by shusvr** [comment]
I believe you want to use [Router::layer](https://docs.rs/axum/latest/axum/struct.Router.html#method.layer) instead.

**Reply by jplatte** [reply]
No, layers applied via `Router::layer` (and `Router::route_layer`) run after routing, but for `NormalizePath` it's important that it runs before routing.

**Comment by simono** [comment, marked as answer]
@agersant I think this should be fixed in axum-test 15.7.0: https://github.com/JosephLenton/axum-test/commit/b7f3ea3bd723970ed4a951e629cdf1ef5787492f
